Add uvmboot tool

kevpar opened the pull request:
I'm trying to ensure two things with the tool:
- Running just uvmboot.exe with no arguments will result in a bog standard UVM, the same thing you'd get if you just call CreateLCOW() with only ID passed as an option.
- uvmboot doesn't hard-code any defaults for parameter values (e.g. as part of urfave/cli argument definitions), so that in case those defaults change in CreateLCOW() down the road, we don't end up with divergent behavior.

To accomplish this, I've added several more things to OptionsLCOW as pointer fields. I know this is kind of gross, but I'm not sure of a better way to do it given the current design. We use IsSet/GlobalIsSet on the cli.Context object to check if parameters are actually set before using their values.

The GCS logs listener has been changed to a generic "output listener", which receives output from the program running in the UVM. It can operate in two modes: either treat output as a Logrus message (parse it and re-log it), or just dump the output to stdout.

Since we normally want stderr from GCS, but for applications just as dmesg we want stdout, there are flags to control whether stdout/stderr are forwarded to HCSShim (with stdout=false and stderr=true defaults).

To avoid cases where the UVM exits before we have received all of its output, I've changed the UVM Wait() function to also wait till the output processing is done. This is signaled via a channel.

A wcow subcommand is not supported yet, but could be added in the future.

jterry75 commented on internal/uvm/create.go:
```diff
@@ -12,6 +12,7 @@ type Options struct {
 	Owner                   string                  // Specifies the owner. Defaults to executable name.
 	Resources               *specs.WindowsResources // Optional resources for the utility VM. Supports Memory.limit and CPU.Count only currently. // TODO consider extending?
 	AdditionHCSDocumentJSON string                  // Optional additional JSON to merge into the HCS document prior
+	UseGuestConnection      *bool                   // Whether the HCS should connect to the UVM's GCS
```
This is LCOW only. Windows without a Guest Connection makes no sense

kevpar replied:
We'd need a guest connection even to test the UVM boot time for WCOW? If so, I'll change it to LCOW only.

jterry75 replied:
I cant think of any way we can get output from Windows without a GuestConnection. We do not control the runtime inside the guest

kevpar replied:
Good to know, I'll fix this up.

kevpar commented:
Currently this log output is always present when running uvmboot with an --exec that terminates (which casues a kernel panic):

```
WARN[0001] Received a callback of an unsupported type    notification-type=13
WARN[0001] Received a callback of an unsupported type    notification-type=6
ERRO[0001] hcsshim::ComputeSystem - End Operation - Error  cid=uvmboot-0 error="Wait uvmboot-0: The virtual machine or container exited unexpectedly." hcs-op="hcsshim::ComputeSystem::Wai
t"
ERRO[0001] hcsshim::ComputeSystem - End Operation - Error  cid=uvmboot-0 error="Terminate uvmboot-0: The virtual machine or container with the specified identifier is not running." hcs-o
p="hcsshim::ComputeSystem::Terminate"
ERRO[0001] Wait uvmboot-0: The virtual machine or container exited unexpectedly.  uvm-id=uvmboot-0
```

For the unsupported callback types, I can just add those to our list of callbacks.

For the other messages, I'm not sure what's best. They're kind of "by design" because HCS doesn't expect the UVM to die on its own, but they're certainly annoying. Users of uvmboot can probably ignore them (and we could add a note in the help output saying they'll be present), or I could make a way to ignore error output from Logrus, or the user could just do "2>nul" when running uvmboot...

Thoughts?

jterry75 commented on internal/uvm/create_lcow.go:
```diff
@@ -22,9 +22,18 @@ import (
+type OutputHandlingType int
+
+const (
+	OutputHandlingTypeLogrus OutputHandlingType = iota
```
What is the difference between these two?

kevpar replied:
I'll add a comment explaining. OutputHandlingTypeLogrus means we parse the output as if it were a Logrus message (this is the existing behavior in master, used to receive messages from GCS). OutputHandlingTypeStdout means we just print it to stdout (e.g. to display dmesg).

jterry75 commented on internal/uvm/create_lcow.go:
Why false by default?

jterry75 replied:
Oh interesting I didn't realize we didn't forward it today.

kevpar replied:
This preserves existing behavior of forwarding stderr from GCS (because it has logrus messages), but not stdout. If GCS doesn't output anything to stdout, changing this wouldn't break anything, but if it does, we'd try to parse it as a logrus message.

jterry75 commented on internal/uvm/create_lcow.go:
Can you specify the defaults in the comments here. Makes it a lot easier to understand

jterry75 commented on internal/uvm/create_lcow.go:
```diff
@@ -168,6 +189,13 @@ func CreateLCOW(opts *OptionsLCOW) (_ *UtilityVM, err error) {
 		},
 	}
 
+	if *opts.UseGuestConnection {
```
You could just write this as `if opts.UseGuestConnection == nil || *opts.UseGuestConnection` and you wouldn't need the fixup above.

kevpar replied:
I liked the organization of having a section at the top to set defaults on options, and then use those values later knowing that they've all been set up properly. If you prefer it this way, I can change it, though.

jterry75 replied:
No I am fine with that. I will likely remove the pointers part later anyways for simplicity

jterry75 commented on internal/uvm/create_lcow.go:
```diff
@@ -261,11 +289,25 @@ func CreateLCOW(opts *OptionsLCOW) (_ *UtilityVM, err error) {
+	initArgs := "/bin/vsockexec"
+
+	if *opts.ForwardStdout {
+		initArgs += fmt.Sprintf(" -o %d", linuxLogVsockPort)
```
Can these really reuse the same port with no issues?

jstarks replied:
Yes, vsockexec is smart enough to just dup the fd in that case.

jterry75 commented on internal/uvm/start.go:
```diff
@@ -58,11 +55,36 @@ func forwardGcsLogs(l net.Listener) {
 }
 
+func printToStdout(c net.Conn) {
+	io.Copy(os.Stdout, c)
```
You should `c.Close()` when this is done

jstarks commented on tools/uvmboot/main.go:
```diff
@@ -0,0 +1,264 @@
+					Usage: "Additional arguments to pass to the kernel",
+				},
+				cli.UintFlag{
+					Name:  rootFSTypeArgName,
```
>rootFSTypeArgName [](start = 12, length = 17)

can we make this param either a string or a bool?

jterry75 commented on internal/uvm/wait.go:
```diff
@@ -1,6 +1,19 @@
-	return uvm.hcsSystem.Wait()
+	err := uvm.hcsSystem.Wait()
+
+	logrus.WithField(logfields.ContainerID, uvm.ID).
```
You meant `logfields.UVMID`

kevpar replied:
I just noticed that I used ContainerID because that's what's used by internal\hcs\system.go. Meaning today we have the begin/end operation log messages logging with cid=<uvm id> for UVMs. We might want to look at fixing this in the future.

kevpar replied:
ERRO[0001] hcsshim::ComputeSystem - End Operation - Error  cid=uvmboot-0 error="Start uvmboot-0: The virtual machine or container exited unexpectedly." hcs-op="hcsshim::ComputeSystem::Start"

jstarks commented on internal/uvm/start.go:
```diff
@@ -58,11 +55,36 @@ func forwardGcsLogs(l net.Listener) {
-		uvm.gcslog = nil
+	if uvm.outputListener != nil {
+		var handler outputHandler
+		switch uvm.outputHandling {
```
>outputHandling [](start = 13, length = 14)

should the options struct just contain a handler so that uvmboot can provide the printToStdout handler itself?

jterry75 commented on internal/uvm/start.go:
```diff
@@ -58,11 +55,36 @@ func forwardGcsLogs(l net.Listener) {
+	defer close(doneChan)
+
+	c, err := l.Accept()
+	l.Close()
```
Do this before `l.Accept()`

jstarks replied:
Close before Accept?

jterry75 replied:
`defer l.Close()`

jstarks replied:
Why though?

jterry75 replied:
I re-read the code not on my phone. This is fine

jterry75 commented on internal/uvm/create.go:
```diff
@@ -35,9 +36,10 @@ func (uvm *UtilityVM) OS() string {
-	if uvm.gcslog != nil {
-		uvm.gcslog.Close()
-		uvm.gcslog = nil
+	if uvm.outputListener != nil {
```
I don't understand how this will ever be true. In `Start()` you set it to `nil` after forwarding it to the process call.

kevpar replied:
I think it's there in case you do Create -> Stop without a Start in between.

jterry75 replied:
Can you add a comment in Stop:
```go
// outputListener will only be nil for a Create -> Stop without a Start. In this case
// we have no goroutine processing output so its safe to close the channel here.
if uvm.outputListener != nil {
  // ...
}
```

jterry75 commented on internal/uvm/wait.go:
```diff
@@ -1,6 +1,19 @@
+
+	logrus.WithField(logfields.ContainerID, uvm.ID).
+		Debug("UVM exited, waiting for output processing to complete")
+	if uvm.outputProcessingDone != nil {
```
Dont you want these in the other order. Wait for logs, wait for uvm to finish terminating?

kevpar replied:
I'm not sure it matters, but I was thinking there might be more output arriving as the VM terminates, so we'd do things in this order to ensure we don't ignore any output.

jterry75 replied:
Ok

jterry75 commented on tools/uvmboot/main.go:
```diff
@@ -0,0 +1,264 @@
+			Name:  debugArgName,
+			Usage: "Enable debug level logging in HCSShim",
+		},
+		cli.BoolFlag{
```
Do you really need this as an opt? What case do we want a guest connection for this tool?

kevpar replied:
I thought there may be a use case for being able to boot a normal UVM and have the logrus output streaming to your console, so you can view it as you launch containers in that UVM. With the current configuration, uvmboot can be used to launch a completely standard UVM as well as a "test" UVM.

jterry75 replied:
This is a really weird scenario. This would work you are correct but it would boot successfully, negotiate with the bridge, and then hang forever waiting for a create container/container exit sequence. To support this requires some more thought I think. Lets remove it for now.

jterry75 commented:
> Currently this log output is always present when running uvmboot with an --exec that terminates (which casues a kernel panic):
> 
> ```
> WARN[0001] Received a callback of an unsupported type    notification-type=13
> WARN[0001] Received a callback of an unsupported type    notification-type=6
> ERRO[0001] hcsshim::ComputeSystem - End Operation - Error  cid=uvmboot-0 error="Wait uvmboot-0: The virtual machine or container exited unexpectedly." hcs-op="hcsshim::ComputeSystem::Wai
> t"
> ERRO[0001] hcsshim::ComputeSystem - End Operation - Error  cid=uvmboot-0 error="Terminate uvmboot-0: The virtual machine or container with the specified identifier is not running." hcs-o
> p="hcsshim::ComputeSystem::Terminate"
> ERRO[0001] Wait uvmboot-0: The virtual machine or container exited unexpectedly.  uvm-id=uvmboot-0
> ```
> For the unsupported callback types, I can just add those to our list of callbacks.
> 
> For the other messages, I'm not sure what's best. They're kind of "by design" because HCS doesn't expect the UVM to die on its own, but they're certainly annoying. Users of uvmboot can probably ignore them (and we could add a note in the help output saying they'll be present), or I could make a way to ignore error output from Logrus, or the user could just do "2>nul" when running uvmboot...
> 
> Thoughts?

For the Wait error. Seems like we could add an overload in this case that "expects" an error when the caller knows that the function will not cause a graceful wait.
For the Terminate error. Seems like this specific error should not be a failure. If you call Terminate on a "terminated" compute system this is a success!

kevpar commented:
> or the Wait error. Seems like we could add an overload in this case that "expects" an error when the caller knows that the function will not cause a graceful wait.
> For the Terminate error. Seems like this specific error should not be a failure. If you call Terminate on a "terminated" compute system this is a success!

I guess I made a comment that can't be replied to as a normal comment thread? I'm not good with GitHub. :(

Wait: I've added a new WaitExpectedError() that doesn't fail if the given error is encountered.
Terminate: I've modified the existing Terminate() function to not return an error if the compute system is already terminated.

With these changes, I can now run uvmboot without any error text!

```
.\uvmboot lcow --exec "uname -r" --fwd-stdout --output-handling stdout --guest-con=false
4.14.35-linuxkit
```

kevpar commented:
I'm not sure the AppVeyor checks are properly running again when I update the branch. I fixed the issue in main.go, but the check is still failing.

```
PS C:\Users\kevpar\Source\go\src\github.com\microsoft\hcsshim\tools\uvmboot> gofmt -s main.go | out-file -enc ascii main2.go
PS C:\Users\kevpar\Source\go\src\github.com\microsoft\hcsshim\tools\uvmboot> get-filehash main.go,main2.go

Algorithm       Hash                                                                   Path
---------       ----                                                                   ----
SHA256          C63F55EDBF1C12F6EBD42E1E7E001C18E3C44BAD4AEFE0B80A9477467F386746       C:\Users\kevpar\Source\go\src\github.com\microsoft\hcsshim\tools\uvmboot\main.go
SHA256          C63F55EDBF1C12F6EBD42E1E7E001C18E3C44BAD4AEFE0B80A9477467F386746       C:\Users\kevpar\Source\go\src\github.com\microsoft\hcsshim\tools\uvmboot\main2.go
```

jterry75 commented on internal/hcs/callback.go:
```diff
@@ -16,11 +16,20 @@ var (
-	hcsNotificationSystemStartCompleted  hcsNotification = 0x00000003
-	hcsNotificationSystemPauseCompleted  hcsNotification = 0x00000004
-	hcsNotificationSystemResumeCompleted hcsNotification = 0x00000005
+	hcsNotificationSystemExited                      hcsNotification = 0x00000001
```
This needs to be a separate PR.

jterry75 commented on internal/hcs/system.go:
```diff
@@ -318,7 +318,7 @@ func (computeSystem *System) Terminate() (err error) {
 	completed = true
 	events := processHcsResult(resultp)
-	if err != nil {
+	if err != nil && err != ErrVmcomputeAlreadyStopped {
```
Would be nice if this was a separate PR for tracking as well

jterry75 commented on internal/hcs/system.go:
```diff
@@ -339,6 +339,20 @@ func (computeSystem *System) Wait() (err error) {
 	return nil
 }
 
+// WaitExpectedError synchronously waits for the compute system to shutdown or terminate, and ignores the passed error if it occurs.
```
NIT: Install ReWrap for VSCode and it will help to leave comments in an easier to read length (80 chars)

jterry75 commented:
> I'm not sure the AppVeyor checks are properly running again when I update the branch. I fixed the issue in main.go, but the check is still failing.
> 
> ```
> PS C:\Users\kevpar\Source\go\src\github.com\microsoft\hcsshim\tools\uvmboot> gofmt -s main.go | out-file -enc ascii main2.go
> PS C:\Users\kevpar\Source\go\src\github.com\microsoft\hcsshim\tools\uvmboot> get-filehash main.go,main2.go
> 
> Algorithm       Hash                                                                   Path
> ---------       ----                                                                   ----
> SHA256          C63F55EDBF1C12F6EBD42E1E7E001C18E3C44BAD4AEFE0B80A9477467F386746       C:\Users\kevpar\Source\go\src\github.com\microsoft\hcsshim\tools\uvmboot\main.go
> SHA256          C63F55EDBF1C12F6EBD42E1E7E001C18E3C44BAD4AEFE0B80A9477467F386746       C:\Users\kevpar\Source\go\src\github.com\microsoft\hcsshim\tools\uvmboot\main2.go
> ```

Just use `gofmt -s -w file.go`
